apps/studio/state/recent-items.ts

```diff
@@ -0,0 +1,167 @@
+// Import necessary functions from valtio for state management
+import { proxy, subscribe } from 'valtio'
+// Import Tab and TabType types for defining recent items
+import { Tab, TabType } from './tabs'
+
+// Define the key for storing recent items in localStorage
+const RECENT_ITEMS_KEY = 'supabase_recent_items'
+// Define the maximum number of recent items to store
+const MAX_RECENT_ITEMS = 8
+
+const getStorageKey = (ref: string) => `${RECENT_ITEMS_KEY}_${ref}`
+
+// Define the structure of a recent item
+export interface RecentItem {
+  id: string // Unique identifier for the item
+  type: TabType // Type of the tab (e.g., sql, table)
+  label: string // Label for the item
+  timestamp: number // Timestamp of when the item was added
+  metadata?: {
+    // Optional metadata associated with the item
+    schema?: string // Schema name (if applicable)
+    name?: string // Name of the item (if applicable)
+    tableId?: number // Table ID (if applicable)
+    sqlId?: string // SQL ID (if applicable)
+  }
+}
+
+// Define the structure of the recent items state
+interface RecentItems {
+  items: RecentItem[] // Each project just has an array of items
+}
+
+interface RecentItemsStore {
+  [key: string]: RecentItems // Removed | undefined since we always initialize it
+}
+
+const defaultState: RecentItems = {
+  items: [],
+}
+
+export const recentItemsStore = proxy<RecentItemsStore>({})
+
+export const getRecentItemsStore = (ref: string | undefined): RecentItems => {
+  if (!ref) return proxy(defaultState)
+  if (!recentItemsStore[ref]) {
+    const stored = localStorage.getItem(getStorageKey(ref))
+    console.debug('[Recent Items] Loading stored items:', stored)
+
+    try {
+      const parsed = stored ? JSON.parse(stored) : defaultState
+      // Validate the shape of the data
+      if (!parsed.items || !Array.isArray(parsed.items)) {
+        console.warn('[Recent Items] Invalid stored data, using default')
+        recentItemsStore[ref] = proxy({ ...defaultState })
+      } else {
+        recentItemsStore[ref] = proxy(parsed)
+      }
+    } catch (error) {
+      console.error('[Recent Items] Failed to parse stored items:', error)
+      recentItemsStore[ref] = proxy({ ...defaultState })
+    }
+  }
+
+  return recentItemsStore[ref]
+}
+
+// Subscribe to changes for each ref and save to localStorage
+if (typeof window !== 'undefined') {
+  subscribe(recentItemsStore, () => {
+    Object.entries(recentItemsStore).forEach(([ref, state]) => {
+      try {
+        localStorage.setItem(getStorageKey(ref), JSON.stringify(state))
+      } catch (error) {
+        console.error('Failed to save recent items to localStorage:', error)
+      }
+    })
+  })
+}
+
+// Function to add a recent item
+export const addRecentItem = (ref: string | undefined, tab: Tab) => {
```

Comment:
Small nit with these stores; our other stores have the functions inline, but here it seems to be following a different pattern. Not a big deal but consistency should not be undervalued!

Reply:
left a comment - will refactor separately as this will involve a rewrite of the tabs and recent-items store 😅